sys/freebsd/sys.txt

```diff
@@ -129,7 +133,16 @@ setitimer(which flags[getitimer_which], new ptr[in, itimerval], old ptr[out, iti
 exit(code intptr)
 wait4(pid pid, status ptr[out, int32, opt], options flags[wait_options], ru ptr[out, rusage, opt])
 
+thr_new(param ptr[in, thr_param], param_size bytesize[param])
+thr_self(lwpid ptr[out, lwpid])
+thr_kill(lwpid lwpid, sig int32)
+thr_kill2(pid pid, lwpid lwpid, sig int32)
```

Comment:
So in general I believe it's not a good idea to enable thr_kill() or thr_kill2(). Right now the fuzzer doesn't run in a sandbox of any kind, such a jail that would restrict visibility of the set of processes, so if it starts killing other processes it'll bring down the VM pretty quickly. We don't fuzz kill(), though that syscall is even more dangerous because kill(-1) sends a signal to all processes.

If we added a jail sandbox mode (Linux has several sandbox types for instance) that might make this more viable.

Reply:
After taking a brief look at executor, I don't think a jail sandbox would be too hard to setup (independently, I'll still punt on these for now). I guess we'd probably need to write a quick MAC module to grant all priv(9) to jail root to maximize coverage, though, and it's not clear where that should live given the security implications.

Reply:
> it's not clear where it's decided what gets ran in a sandbox

You mean from the OS point of view?... of from executor point of view? Is it something about executor that is not clear?

Overall a good sandbox should be very useful. On Linux we use it to isolate the test process in terms of what it can reach and resource consumption and prevent it from doing unintentional bad things.

Reply:
> > it's not clear where it's decided what gets ran in a sandbox
> 
> You mean from the OS point of view?... of from executor point of view? Is it something about executor that is not clear?

Sorry, from the executor's point of view. I think this became clear as I read on, that it's configurable per-syzkaller-instance rather than "fuzzing involving this set of syscalls requires a sandbox."

> Overall a good sandbox should be very useful. On Linux we use it to isolate the test process in terms of what it can reach and resource consumption and prevent it from doing unintentional bad things.

Yeah, so jails would be a solid choice here, we'll just need to work out some logistics. Jail root is pretty restricted compared to your standard super-user, a number of syscalls will get rejected with EPERM in stock FreeBSD pretty early on in execution. I've got some WIP that might also alleviate this by letting us lift the jail's priv set up to the system root, so we still get the benefits of, e.g., limited visibility and affect on pid space, while not restricting the coverage we can achieve.

Reply:
> that it's configurable per-syzkaller-instance rather than "fuzzing involving this set of syscalls requires a sandbox."

Yes, sandbox is chosen once in syz-manager config and then all test programs run in this sandbox.

If you are going to implement this, just a bit of experience with our Linux sandboxes. In hindsight having lots of sandboxing modes (root/namespace) doesn't make lots of sense. Having a single sandboxing mode that is the "right" balance between permitting/not permitting things would be enough. So I would suggest to just change the current default sandboxing mode for FreeBSD (unless there is a very clear and explicit use case for more modes).